scipy/stats/tests/test_stats.py

```diff
@@ -7249,6 +7294,88 @@ def test_inf_values(self):
+        distance_1 = stats.wasserstein_distance(uv_1, vv_1, uw_1, vw_1)
+        uv_2, vv_2 = [[-1], [0], [4], [8]], [[-1], [3], [2], [0]]
+        uw_2, vw_2 = [1, 3, 5, 7], [2, 4, 6, 8]
+        distance_2 = stats.wasserstein_distance(uv_2, vv_2, uw_2, vw_2)
+        assert_almost_equal(distance_1, distance_2)
+
+    @pytest.mark.parametrize('input_dim', (2, 3, 7, 15, 60))
+    @pytest.mark.parametrize('data_volumn', ((3, 5), (12, 16), (20, 20),
+                                             (1, 200), (15, 15)))
+    def test_shift_nD(self, input_dim, data_volumn):
+        # Moving across n-dimensional space does not change the outcome.
+        rng = np.random.default_rng(734825773)
+        uu = rng.random((data_volumn[0], input_dim))
+        vv = rng.random((data_volumn[1], input_dim))
+        expected = stats.wasserstein_distance(uu, vv)
+        shift = rng.random(input_dim)
+        uu_shift, vv_shift = uu + shift, vv + shift
+        shift_result = stats.wasserstein_distance(uu_shift, vv_shift)
+        assert_almost_equal(expected, shift_result)
+
+    @pytest.mark.parametrize('input_dim', (3, 4, 5, 22))
+    @pytest.mark.parametrize('data_volumn', ((6, 3), (10, 10), (30, 8),
+                                             (6, 150), (20, 14)))
+    def test_rotation_reflection_nD(self, input_dim, data_volumn):
+        # Rotating or reflecting across across n-dimensional
+        # space does not change the outcome.
+        rng = np.random.default_rng(734826773)
+
+        uu = rng.random((data_volumn[0], input_dim))
+        vv = rng.random((data_volumn[1], input_dim))
+        expected = stats.wasserstein_distance(uu, vv)
+        diag_value = rng.choice([-1, 1], size=input_dim)
+        # random reflection matrix
+        reflection_matrix = np.diag(diag_value)
+        combinations = []
+        for i in range(input_dim - 1):
+            for j in range(i + 1, input_dim):
+                combinations.append((i, j))
+
+        def rotation(index):
+            # generate random nD rotation matrix
+            # rotation is not uniformly distributed on unit sphere!
+            index_i, index_j = index
+            rotation_matrix = np.identity(input_dim)
+
+            rotation_angle = np.pi * rng.uniform(0, 1)
+            rotation_matrix[index_i, index_i] = np.cos(rotation_angle)
+            rotation_matrix[index_i, index_j] = - np.sin(rotation_angle)
+            rotation_matrix[index_j, index_i] = np.sin(rotation_angle)
+            rotation_matrix[index_j, index_j] = np.cos(rotation_angle)
+            return rotation_matrix
+        rotations = [rotation(index) for index in combinations]
+        rotation_matrix = np.linalg.multi_dot(rotations)
+        rotation_reflection = reflection_matrix @ rotation_matrix
+        uu_rr, vv_rr = uu @ rotation_reflection, vv @ rotation_reflection
+
+        shift_result = stats.wasserstein_distance(uu_rr, vv_rr)
+        assert_almost_equal(expected, shift_result)
```

Comment:
These are great, but let's combine them. I think the random reflection/rotation matrix can be created by drawing a matrix from the orthogonal group.

```python3
rng = np.random.default_rng(3178853690258791578)
nu, nv, ndim = 10, 15, 5  # you can parameterize this part

u_values = rng.random(size=(nu, ndim))
v_values = rng.random(size=(nv, ndim))
u_weights = rng.random(size=nu)
v_weights = rng.random(size=nv)
ref = wasserstein_distance(u_values, v_values, u_weights, v_weights)

dist = stats.ortho_group(ndim)
transform = dist.rvs(random_state=rng)
shift = rng.random(size=ndim)
res = wasserstein_distance(u_values @ transform + shift,
                           v_values @ transform + shift,
                           u_weights, v_weights)
assert_allclose(res, ref)
```